device/cuda/src/clusterization/clusterization_algorithm.cu

```diff
@@ -0,0 +1,305 @@
+        m_mr.get());
+    copy.setup(sparse_ccl_indices_buff);
+
+    /*
+     * cl_per_module_prefix_buff is a vector buffer with numbers of found
+     * clusters in each module. Later it will be transformed into prefix sum
+     * vector (hence the name). The logic is the following. After
+     * cluster_finding_kernel, the buffer will contain cluster sizes e.i.
+     *
+     * cluster sizes: | 1 | 12 | 5 | 102 | 42 | ... - cl_per_module_prefix_buff
+     * module index:  | 0 |  1 | 2 |  3  |  4 | ...
+     *
+     * Now, we copy those cluster sizes to the host and make a duplicate vector
+     * of them. So, we are left with cl_per_module_prefix_host, and
+     * clusters_per_module_host - which are the same. Now, we procede to
+     * modifying the cl_per_module_prefix_host to actually resemble its name
+     * i.e.
+     *
+     * We do std::inclusive_scan on it, which will result in a prefix sum
+     * vector:
+     *
+     * cl_per_module_prefix_host: | 1 | 13 | 18 | 120 | 162 | ...
+     *
+     * Then, we copy this vector into the previous cl_per_module_prefix_buff.
+     * In this way, we don't need to allocate the memory on the device twice.
+     *
+     * Now, the monotonic prefix sum buffer - cl_per_module_prefix_buff, will
+     * allow us to insert the clusters at the correct position inside the
+     * kernel. The remaining host vector - clusters_per_module_host, will be
+     * needed to allocate memory for other buffers later in the code.
+     */
+    vecmem::data::vector_buffer<std::size_t> cl_per_module_prefix_buff(
+        num_modules, m_mr.get());
+    copy.setup(cl_per_module_prefix_buff);
+
+    // Create views to pass to cluster finding kernel
+    const cell_container_types::const_view cells_view(cells_data);
+    vecmem::data::jagged_vector_view<unsigned int> sparse_ccl_indices_view =
+        sparse_ccl_indices_buff;
+    vecmem::data::vector_view<std::size_t> cl_per_module_prefix_view =
+        cl_per_module_prefix_buff;
+
+    // Calculating grid size for cluster finding kernel
+    std::size_t blocksPerGrid =
+        (num_modules + threadsPerBlock - 1) / threadsPerBlock;
+
+    // Invoke find clusters that will call cluster finding kernel
+    kernels::find_clusters<<<blocksPerGrid, threadsPerBlock>>>(
+        cells_view, sparse_ccl_indices_view, cl_per_module_prefix_view);
+
+    // Get the prefix sum of the cells and copy it to the device buffer
+    const device::prefix_sum_t cells_prefix_sum =
+        device::get_prefix_sum(cell_sizes, m_mr.get());
+    vecmem::data::vector_buffer<device::prefix_sum_element_t>
+        cells_prefix_sum_buff(cells_prefix_sum.size(), m_mr.get());
+    copy.setup(cells_prefix_sum_buff);
+    copy(vecmem::get_data(cells_prefix_sum), cells_prefix_sum_buff,
+         vecmem::copy::type::copy_type::host_to_device);
+
+    // Wait here for the cluster_finding kernel to finish
+    CUDA_ERROR_CHECK(cudaGetLastError());
```

Comment:
I **love** that you put the `cudaDeviceSynchronize()` call after the prefix sum creation. But I think `cudaGetLastError()` should still be put right after the kernel launch. Since it is meant to detect a problem with the kernel launch itself. `cudaDeviceSynchronize()` on the other hand would communicate errors that happened during the running of the kernel.

Reply:
I might not have noticed that prefix sum creation is independent and can be done parallelly if not for what Konrad had done with Sycl. :sweat_smile:
Noted and I made the suggested change.